ee/WFLY-1160-Provide_ability_to_configure_a_global_directory_which_puts_the_contents_in_the_classpath.adoc

```diff
@@ -0,0 +1,94 @@
+= Provide ability to easily apply certain JBoss module libraries to all deployments running in a server
+:author:            Yeray Borges
+:email:             yborgess@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+Wildfly users can share libraries across all deployed applications without having to modify the packaging of their applications creating a custom module and making it global using the EE subsystem:
+
+  [standalone@localhost:9990 /] module add --name=com.test --absolute-resources=/Users/me/libs:/Users/me/libs/my-utils1.jar:/Users/me/libs/my-utils2.jar
+  [standalone@localhost:9990 /] /subsystem=ee:list-add(name=global-modules,value={name=com.test})
+
+This configuration implies the following:
+
+  * Users have to specify all required resource roots using --absolute-resources argument.
+  * It is necessary to modify the generated module.xml, or at least re-create the global module if we want to add/remove any additional resource-root for this global module.
+  * This configuration allows any customization of the generated module.xml, making it very flexible to cover broader cases, for example, it is possible to declare module dependencies, mark them as exported, declare the main class and so on.
+  * This command is not appropriate for use in a managed domain or when connecting to the management CLI remotely. Modules should be added and removed manually in a production environment.
+
+This RFE is opened to have an alternative new operation to address simpler cases, where it is not required any additional configuration beyond on the default module.xml configurations, and supported in production environments and domain mode.
+
+The simplest way to accomplished this is to use the EE subsystem to configure a path to the shared folder. An example of this configuration could be:
+
+  [standalone@localhost:9990 /] /subsystem=ee:list-add(name=global-folders,value={name=my-common-libs, path=/Users/me/libs})
+
+  [domain@localhost:9990 /] /profile=default/subsystem=ee:list-add(name=global-folders,value={name=my-common-libs, path=/Users/me/libs})
+
+The name and path attributes will be mandatory.
+Internally, this operation will create a module with the name `my-common-libs` that will have as resource-roots the folder `/Users/me/libs/` and all the jars files included in this folder. Subdirectories will not be scanned, but if it is required, users can create a new global-folder using the subdirectory path.
```

Comment:
This sentence suggests the functionality work only locally, then only in standalone mode.
It is not clear if functionality work in standalone mode or domain mode.
Can you explain it?

Reply:
My initial idea regarding domain mode is we will able to have a local folder path relative to the host controller file system. 

We will be configuring the EE subsystem with a path. The EE subsystem in domain mode is configured under a profile. Each profile configuration is transferred to each slave HC, so each HC will see just a path configuration which should be local to its filesystem.

That means we are not going to transfer all the available jar files from the DC to the slaves. Is the user responsibility to have a folder available on each host where an HC instance is running. If an HC is not managing servers in a server group affected by the profile which was configured, then this requirement should not be necessary, since there weren't deployments running with this configuration.

Does it make sense to you? I'm still testing if this is possible and what could be the best way to give support in domain mode for this feature.

Reply:
Does it work remotely in the standalone mode? Or do you plan it?

Do you expect the customer will be responsible for distribution jars in the shared library to the server and all slaves manually?

If it will be distributed only name and path. What about path collision between Linux and Windows?
Some nodes can be running on Windows and other nodes can be running on Linux.

Reply:
> Does it work remotely in the standalone mode? Or do you plan it?

Yes, it should work remotely too, it is the server jvm responsibility to check the paths internally, not the client jvm. I could be giving a vague response, please, would you mind to be more specific about your concern? Why do you think it couldn't work if you are configuring the server remotely? What exactly do you think it couldn't work?

> Do you expect the customer will be responsible for distribution jars in the shared library to the server and all slaves manually?

For a standalone mode, they should have in the shared folder the jars they want to include as a dependency of all the deployed applications. In domain mode, the jar should be accesible from the host controller file system too. It is the same if a customer wants to add a new module, they would have to create a module.xml with a resource root pointing to a jar file accessible from the server (standalone mode) or host controller (domain mode) file system.

> If it will be distributed only name and path. What about path collision between Linux and Windows?

It will work like any other operation in domain mode that it is using path and relative-to attributes (we can have paths with a similar name but different values per HC, they can use expressions ...)

Reply:
> > Does it work remotely in the standalone mode? Or do you plan it?
> 
> Yes, it should work remotely too, it is the server jvm responsibility to check the paths internally, not the client jvm. I could be giving a vague response, please, would you mind being more specific about your concern? Why do you think it couldn't work if you are configuring the server remotely? What exactly do you think it couldn't work?

My question was be if can be jar uploaded to the server via CLI.
You've already answered my question, It will don't upload jars, thank you.